HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+the patterns at the time they install it.
+1. `CALLER_PAYS` - The payer of the transaction that triggers the hook pays for its upfront gas cost. They receive the
+normal refund for unused gas.
+2. `CALLER_PAYS_ON_REVERT` - The payer of the triggering transaction again pays for the upfront gas cost, but receives
+a _full refund_ if the hook succeeds (in sense appropriate to the extension point). On success, a designated account
+that signed the lambda's installation pays for the gas actually consumed.
+
+Regardless of the charging pattern, a triggering transaction can always set an explicit gas limit for an EVM hook's
+execution, just as with a contract call. If no explicit limit is set on the transaction, the networks checks if the
+EVM was installed with a default gas limit. If there is no default limit for the hook, the network uses a global
+default, for example `hooks.evm.defaultGasLimit=25_000`.
+
+We propose the same gas price for EVM hook execution as for other contract operations. However, unlike contract calls,
+which are charged purely in gas, hook executions are already "gated" by the fee of their triggering transaction.
+So it makes sense to waive the intrinsic gas cost of their execution. We propose adding two more properties,
+
+```
+hooks.evm.pureIntrinsicGasCost=0
+hooks.evm.lambdaIntrinsicGasCost=0
+```
+
+### Throttling
+
+We propose EVM hooks be subject to the same gas throttle as top-level contract calls. Specifically, when an EVM hook
+executes, its initial EVM sender address is the payer of the referencing transaction. If this payer is throttle exempt,
+no throttles are applied. Otherwise, if the network is at capacity for gas usage, EVM hook execution will be throttled
+on that basis and the triggering transaction fail with a status of `EVM_HOOK_EXECUTION_THROTTLED`.
+
+### EVM hook execution
+
+The first message frame for the EVM transaction that executes an EVM hook will be a logical `DELEGATECALL` from the
+**hook system contract address `0x16d`** to the hook's implementing `ContractID`, sent by the payer of the upfront gas
+cost for the hook's execution.
+
+As a concrete example, suppose account `0.0.A` installs an EVM lambda hook for the `ACCOUNT_ALLOWANCE_HOOK` extension
+point at index `1`. The hook's implementing contract is `0.0.H` with EVM address `0xab...cd`. Now `0.0.B` with EVM
+address `0x01...23` sends a `CryptoTransfer` transaction that references the hook `0.0.A.1` with gas limit `100_000`.
+
+The network will construct an initial EVM message frame with,
+1. `sender` address `0x01...23`;
+2. `receiver` address `0x16d`;
+3. `contract` address `0xab...cd` (hence the source of the executing bytecode);
+4. Storage of the `0.0.A.1` lambda EVM hook; and,
+5. Gas remaining of `100_000` (no intrinsic gas cost).
+
+Since this hook is a lambda, it can then proceed with arbitrary EVM behaviors, including calls to external oracles,
```

Comment:
Contracts can't call to external oracles, this should be removed.

Reply:
Fixed.